### PR TITLE
[ENH] Node2vec node embeddings

### PR DESCRIPTION
**Pull request by matejklemen**
##### Issue
Introduces a new widget (Network embeddings) and a method for embedding nodes into vectors, [node2vec](https://arxiv.org/abs/1607.00653).

The method (very briefly described) works in 2 steps:
1. Obtain sequences of random walks across the network. These walks are guided by two parameters, **p** and **q**.
2. Run word2vec on these sequences to obtain embeddings.

##### Known TODOs
The method and widget already work, but some things need to be improved. 
- [X] additional testing to verify node2vec works properly
- [ ] at least a basic example of use
- [ ] proper icon (currently it uses the same icon as `Network File`)

I'd appreciate any comments regarding the widget.

##### Includes
- [X] Code changes
- [X] Tests
- [ ] Documentation

### CODE REVIEW

**Comment by ajdapretnar** [comment]
As for icon, @larazupan is the best one. :)
I'll work on the documentation.
I guess then @janezd (and myself perhaps) should come up with a basic example.

Big thanks for this, I think it will be very useful!

**Comment by ajdapretnar** [comment]
Also, if you ever feel inspired, Text add-on would also need something like this! 😅

**Comment by ajdapretnar** [comment]
While I think it is nice you added the 'Computation in progress...' info, a more standard way in Orange is to implement a progress bar.

See a tutorial for [progress bar](https://github.com/biolab/orange3/blob/9c539ab173882d237dce50a99908b3fd21e81a6d/doc/development/source/tutorial-utilities.rst) and the [source code](https://github.com/biolab/orange3/blob/37e7b0232760ec39b8206a578f63449fdd004615/Orange/widgets/utils/progressbar.py).

**Comment by matejklemen** [comment]
I've removed the info message and added a progress bar. It's not evenly spaced out, because it's only getting updated during word2vec (step 2) and not during the random walks (step 1).

**Comment by ajdapretnar** [comment]
Don't worry, most progress bars in Orange are not evenly spaced out - most cannot be, because different computations take different amount of time.

**Comment by PrimozGodec** [comment]
Is this one still waiting for the icon?

**Comment by matejklemen** [comment]
Yes, although it's also still waiting for more of my attention :smile: 

I wanna create a nice example (at least a proof of concept) to show the functionality of node2vec and also test the widget some more when doing that. But I haven't found the time so far

**Comment by matejklemen** [comment]
Apologies for taking an absolute eternity to get back to this :sweat_smile: 

I've had a bit of time to do some more testing and I've gathered two networks (Cora and Citeseer), on which I compared the accuracy of reference node2vec implementation and this one. The results seem to roughly match:   

| Settings | Reference | Orange |
| -------- | --------------- | ------------ |
| Cora, p=0.7, q=0.7, num_walks=10, walk_length=100, size=128, window=5, iter=10, C=0.01 | 0.789 | 0.789 |
| CiteSeer, p=0.7, q=0.1, num_walks=10, walk_length=100, size=128, window=5, iter=10, C=0.5 | 0.603 | 0.596 |

I'm attaching the [script and data](https://github.com/biolab/orange3-network/files/5226609/n2v_data.zip) here in case somebody wishes to double check it.
The script should be run inside the src/ folder of reference node2vec implementation - **some `print`s might need to be fixed in the reference implementation** in order for it to work.
```
$ git clone https://github.com/aditya-grover/node2vec
$ cd node2vec/src
$ python run_reference_node2vec.py
```

The Orange equivalent of this is:
Screenshot from 2020-09-15 19-17-59

Bottom line is: I'm now more certain that the implementation works correctly.

**Comment by janezd** [comment]
Great!

I pushed two additional commits.

- I suppose your `__main__` is copied from some other widget's code. We now have a simpler utility function for this.
- The second commit simplifies the user interface. I removed the "method" combo box and a redundant label, and I right-aligned the spins. I apologize for making this kind of changes myself instead of asking you first, but I wanted to try what it would look like ... and by trying, it's already done. :) Feel free to reject this commit if you don't like it.

I'm not an expert for threads. Could somebody (@PrimozGodec? @VesnaT?) take a look if they're OK?

**Comment by matejklemen** [comment]
> I removed the "method" combo box  

Fair enough, I added this as a placeholder in case there will be multiple embedding methods some day, but I suppose it makes more sense to add that later.  
The only problem I could see with removing the method name is that the widget name (`Network embeddings`) is very general and it might not be completely clear to the user that node2vec is being used

I pushed another commit that "fixes" the exception which appears when you run the embeddings widget, then delete it before it's done. The problem was that gensim callback tried to update the progress bar on a deleted object. I've made it so that the widget object whose progress bar the callback is updating, gets set to `None` when the widget gets deleted.  
Doesn't feel like the right way to fix this, but I didn't find a way to remove callbacks.

**Comment by ajdapretnar** [comment]
I'd love to merge this, if possible. @janezd or @matejklemen if one of you could resolve conflicts in travis YAML, I can double-test and merge. Thanks!

**Comment by ajdapretnar** [comment]
I will merge this as is, because it is nice and it works. In the future, it would be nice to add the option to cancel the embedding, as in other embedding widgets. Sometimes the data is large and it takes a long time. Changing one's mind should be an option. 😁
